microprofile/WFLY-12590_Integrate_MicroProfile_Fault_Tolerance_2.0.adoc

```diff
@@ -0,0 +1,102 @@
+[ ] Engineering
+
+[x] QE
+
+=== Affected Projects or Components
+
+* The specification repository resides at official Eclipse organization on GitHub in https://github.com/eclipse/microprofile-fault-tolerance[eclipse/microprofile-fault-tolerance] repository.
+* The implementation of Eclipse MicroProfile Fault Tolerance will be provided by SmallRye project https://github.com/smallrye/smallrye-fault-tolerance[smallrye/smallrye-fault-tolerance].
+* The underlying library used by _current_ SmallRye implementation is Hystrix which resides in official Netflix GitHub organization https://github.com/Netflix/Hystrix[Netflix/Hystrix].
+
+=== Other Interested Projects
+
+Not interested in our integration per se, but projects that have already integrated MP FT:
+
+* https://thorntail.io/[Thorntail] has integrated MicroProfile Fault Tolerance as a https://github.com/thorntail/thorntail/tree/master/fractions/microprofile/microprofile-fault-tolerance[fraction].
+* https://quarkus.io/[Quarkus] has integrated MicroProfile Fault Tolerance as an https://github.com/quarkusio/quarkus/tree/master/extensions/smallrye-fault-tolerance[extension].
+
+== Requirements
+
+* Support Eclipse MicroProfile Fault Tolerance version 2.0 for all applications deployed in WildFly.
+* [WILL BE CLARIFIED] The extension is either added to all configurations provided by WildFly (standalone and domain mode)
+  or provided as a layer in a separate build profile.
+* Upstream Eclipse MicroProfile Fault Tolerance TCK test suite will be added to `microprofile-tck` WildFly upstream test suite.
+
+== Nice-to-Have Requirements
+
+* [WILL BE CLARIFIED] Create a MicroProfile Fault Tolerance https://github.com/wildfly/quickstart[quickstart].
+
+Items that are currently only being taken into consideration with next version of this feature:
+
+* The underlying library used by SmallRye FT implementation is Hystrix which is currently in "maintenance" mode and a final
+  release of the project (version 1.5.18) has been released. The discussion about replacing the library is ongoing.
+* WildFly's standard management model will be configuration-less
+** all configuration is done through MicroProfile Config specification
+** given the maintenance problem above it is not desirable to expose a management model which will change in near future as concepts are not cleanly mappable to a different underlying library
+* Support for plurality of isolated deployments.
+
+== Non-Requirements
+
+* N/A
+
+== Implementation Plan
+
+* Add dependencies on `smallrye-fault-tolerance` and `microprofile-fault-tolerance-api` artifacts.
+* Add a new `microprofile-fault-tolerance-smallrye` extension to provide Fault Tolerance APIs support for deployments.
+* Add configuration-less MicroProfile Fault Tolerance subsystem.
+* Support configuration of Hystrix from MicroProfile Config specifications. The configuration is queried for Hystrix properties which are then applied through Archaius configuration mechanism.
```

Comment:
All Hystrix configuration properties or just subset of properties is going to be enabled?

Motivation: Some of the properties may not be relevant for application server use-case. So we have smaller area to cover if we can limit enabled properties.

Reply:
The management API of the subsystem (whether implemented via MP Config or otherwise) needs to be spelled out in this doc.

Reply:
> All Hystrix configuration properties or just subset of properties is going to be enabled?

I don't see an added value in restricting properties that are applied moreover not sure if thats even viable. Neither of the projects integrating this impose any limits.

> The management API of the subsystem (whether implemented via MP Config or otherwise) needs to be spelled out in this doc. I'm not talking about the spec defined MP Config properties.

@bstansberry I am not sure what you mean here – could you rephrase your comment? There will be no configuration in the subsystem.

Reply:
@rhusar WildFly provides a configuration contract. For these MP specs we are moving toward using MP Config to implement configuration but that doesn't mean it's all uncontrolled and we have random sets of config properties that change from release to release. At a minimum we need to document what we support and note that things may change.

By "I'm not talking about the spec defined MP Config properties" I mean that if the MP FT spec says you use config property mp.ft.foo to configure your level of fooness then there's no need to document that here; it's documented in the spec. It's the custom configuration stuff that needs to be documented.

Re: enabling properties, understood that it may not be possible to turn things off, but if something is not supported that can be documented.

Reply:
I have added a warning that these might change in the future in the documentation. Also, the properties that we expect people to tune are described in the docs.